README.md
RawSourcePlus
=============

### Loading raw video data from files

for Avisynth+ r2150 or greater\
This filter is only for Avisynth+MT. Avisynth2.6 is not supported.

#### requirements

- Windows Vista sp2 or later
- Avisynth+ r2150 or greater
- Visual C++ Redistributable Packages for Visual Studio 2019

#### How to use

```
RawSourcePlus (string "file", int "width", int "height", string "pixel\_type", int "fpsnum", int "fpsden", string "index", bool "show", int "sarnum", int "sarden")
```

RawSourcePlus opens a video file which contains 8bit, 9bit, 10bit, 16bit or float YUV444, YUV422, YUV411, YUV420, Gray, or RGB video data.

There are three ways how the positions of the video frame data are calculated:
- a YUV4MPEG2-header is found, width/height/framerate/pixeltype/fieldorder is set according to the header data. Only fixed-length FRAME headers without 'm'tag are supported.
- width, height and pixel\_type is given as arguments. Then the positions are calculated assuming that only video data is in the file.
- an "index" string (or file) is given together with width, height, pixel\_type. The index string can contain all or partial positions of the frames.

Supported pixel_types are:
- RGB, RGBA, BGR, BGRA, ARGB, ABGR (interleaved RGB without subsampling, resulting in AviSynth's RGB24 or RGB32)
- YUY2(YUYV), YVYU, UYVY, VYUY (interleaved horizontally subsampled resulting in AviSynth's YUY2)
- YV24, YUV444P8 (planar 8bit without subsampling, resulting in AviSynth's YV24)
- YV16, YUV422P8 (planar 8bit horizontally subsampled, it is converted to AviSynth's YV16)
- YV12, I420, IYUV, YUV420P8 (planar 8bit horizontally and vertically subsampled resulting in AviSynth's YV12)
- YV411, YUV411P8, Y41B (planar horizontally subsampled, it is converted to AviSynth's YV411)
- NV12, NV21 (chroma interleaved 8bit horizontally and vertically subsampled, those are converted to AviSynth's YV12)
- Y8, GRAY8 (luma only resulting in AviSynth's Y8)
- RGB48, RGBA64, BGR48, BGRA64, ARGB64, ABGR64 (interleaved RGB without subsampling, resulting in AviSynth+'s RGB48 or RGB64)
- YUV444P9, YUV444P10, YUV444P12, YUV444P14, YUV444P16, YUV444PS (planar 9\~16bit or float without subsampling, resulting in AviSynth+'s YUV444P10/12/14/16/S)
- YUV422P9, YUV422P10, YUV422P12, YUV422P14, YUV422P16, YUV422PS (planar 9\~16bit or float horizontally subsampled, resulting in AviSynth+'s YUV422P10/12/14/16/S)
- YUV420P9, YUV420P10, YUV420P12, YUV420P14, YUV420P16, YUV420PS (planar 9\~16bit or float horizontally and vertically subsampled resulting in AviSynth+'s YUV420P10/12/14/16/S)
- P210, P216 (chroma interleaved 16bit horizontally subsampled, those are converted to AviSynth+'s YUV422P16)
- P010, P016 (chroma interleaved 16bit horizontally and vertically subsampled, those are converted to AviSynth+'s YUV420P16)
- Y10, Y12, Y14, Y16, Y32, GREY16, GREYS (luma only 10\~16bit or float resulting in AviSynth+'s Y10/12/14/16/32)
- GBRP, GBRP10, GBRP12, GBRP14, GBRP16, GBRPS (planar 8\~16bit or float RGB without subsampling, resulting in Avisynth+'s RGBP/10/12/14/16/S)
- GBRAP, GBRAP10, GBRAP12, GBRAP14, GBRAP16, GBRAPS (planar 8\~16bit or float RGB with alpha, resulting in Avisynth+'s RGBAP/10/12/14/16/S)
- YUVA444, YUVA444P10, YUVA444P12, YUVA444P14, YUVA444P16, YUVA444PS (planar 8\~16bit or float YUV444 with alpha, resulting in AviSynth+'s YUVA444/P10/P12/P14/P16/PS)
- YUVA422, YUVA422P10, YUVA422P12, YUVA422P14, YUVA422P16, YUVA422PS (planar 8\~16bit or float YUV422 with alpha, resulting in AviSynth+'s YUVA422/P10/P12/P14/P16/PS)
- YUVA420, YUVA420P10, YUVA420P12, YUVA420P14, YUVA420P16, YUVA420PS (planar 8\~16bit or float YUV420 with alpha, resulting in AviSynth+'s YUVA420/P10/P12/P14/P16/PS)

The default value of framerate is 25fps, you can change it with specified 'fpsnum' and 'fpsden' if you need (e.g. for NTSC-material).

With show=true the actually used byteposition for that frame is displayed, followed by:\
K = position given in index is used\
D = position by adding current delta is used\
B = position by adding currend big\_delta is used

#### Some simple examples:

```
RawSourcePlus("d:\yuv4mpeg.y4m")  #this assumes there is a valid YUV4MPEG2-header inside.
# If not, default values are used: width=720, height=576, pixel_type="YUY2".
```

```
RawSourcePlus("d:\src6_625.raw",720,576,"BGRA") # really a raw file
# If you are unsure about the pixel_type, simply try out all possible values.
```

#### Using an index-string:

You can enter the byte positions of the video frames directly.

```
RawSourcePlus("d:\yuv.mov",720,576,"UYVY", index="0:192512 1:1021952 25:21120512 50:42048512 75:62976512")
```

This is useful if it's not really raw video, but e.g. uncompressed MOV files or a file with some kind of header.\
It will work whenever the spacing of the frames is fixed or has at least two fixed intervalls (e.g. audio data interleaved with the video every 25th frame).\
You enter pairs of `framenumber:byteposition`.\
Internally there are two step values (for the byte positions): delta and big\_delta.\
delta is stored everytime when two adjacent framenumbers are given, the default value is width\*height\*bytes\_per\_pixel.\
big\_delta is stored, when three framenumbers with the same two intervals are given. The default value is 0 (meaning there is no useful big\_delta present).\
If those conditions are not met, the internal values of delta and big\_delta is not updated, only the given bytepositions in the index are used.\
big\_delta is reset to 0 if the resulting position would be behind the given one (see beyond).

Here are some possible cases:

| Case | Description |
| --- | --- |
| 0:    0 | frame 0 starts at byte 0, step to frame 1 is default = width*height*bytes_per_pixel |
| 0:10000 | frame 0 starts at 10000 |
| 0:  5000<br>1: 15000<br><br><br><br>25:290000<br><br><br>50:590000<br><br><br><br>75:890000<br><br><br> | frame 0 at 5000<br>frame 1 at 15000 (delta=10000)<br>frame 2 at 25000<br>...<br>frame 25 at 290000 (using entry instead of delta which would be at 255000)<br>frame 26 at 300000 (still using delta)<br>...<br>frame 50 at 590000 (using entry instead of delta)<br>>> because 25...50 = 50...75 now big_delta is set to 300000 (590000-290000)<br>frame 51 at 600000 (still using delta)<br>...<br>frame 75 at 890000 (using entry which is the same as using big_delta)<br>...<br>frame 100 at 1190000 (using big_delta)<br>frame 101 at 1200000 (using delta)...frame 125 at 1490000 (using big_delta) |
| 0:  5000<br>1: 15000<br>25:290000<br>50:590000<br>75:890000<br><br>100:950000<br><br><br><br> | the same as in the previous example<br><br><br><br>frame 75 at 890000<br>>> because 890000+300000 > 950000 now big_delta is reset to 0.<br>frame 100 at 950000<br>frame 101 at 960000 (using delta)<br>...<br>frame 125 at 1200000 (there is NO big_delta) |

The index string is treated as a filename, if there is an "." inside. The data is then read from that file, line breaks don't matter.

#### note:

This filter is automatically registerd as MT\_SERIALIZED.\
You don't have to set it yourself.

#### Exported variables:

FFSAR_NUM, FFSAR_DEN, FFSAR.

#### Building

##### Windows

Use solution files.

##### Linux

###### Requirements

- Git
- C++11 compiler
- CMake >= 3.16

```
git clone https://github.com/Asd-g/RawSource_2.6x && \
cd RawSource_2.6x && \
mkdir build && \
cd build && \

cmake ..
make -j$(nproc)
sudo make install
```

#### LICENSE:

Same as the original plugin.


CHANGELOG.md
#### original author:Ernst Peché, 2005-10-13

#### modified by Oka Motofumi, 2011-06-14

`Version 2016-07-07 - Modified RawSource26 to Avisynth+ plugin.`

`Version 2016-08-14 - Update for Avisynth+ r2150 or later.`

`Version 2020-05-13 - Fixed 10/12/14-bit input, v8 interface.`

`Version 0.0.1 - Set few frame properties.`

```
Version 1.0.0
- Used reserved frame property names.
- Added parameters sarnum and sarden.
- Set frame properties _SARNum and _SARDen.
- Set variables FFSAR_NUM, FFSAR_DEN, FFSAR.
```

```
Version 1.0.1
- Fixed frame properties _DurationNum and _DurationDen.
```

src/rawsource26.cpp
/*
 RawSourcePlus - reads raw video data files

    Author: Oka Motofumi (chikuzen.mo at gmail dot com)

    This program is rewriting of RawSource.dll(original author is Ernst Pech)
    for Avisynth+.
*/


#ifdef _WIN32
#include <io.h>
#else
#define _FILE_OFFSET_BITS     64
#include "win_import_min.h"
#include <sys/stat.h>
#endif
#include <fcntl.h>
#include <malloc.h>
#include <algorithm>
#include <cinttypes>
#include <string>
#include <tuple>
#include <unordered_map>
#include "common.h"



class RawSource : public IClip {

    VideoInfo vi;
    int fileHandle;
    int64_t fileSize;
    int order[4];
    int col_count;
    bool show;
    bool has_at_least_v8;
    int sar_num;
    int sar_den;

    uint8_t* rawbuf;
    i_struct* index;

    void setProcess(const char* pix_type);

    void(__stdcall *writeDestFrame)(
        int fd, PVideoFrame& dst, uint8_t* buff, int* order, int count,
        ise_t* env);

public:
    RawSource(const char* source, const int width, const int height,
              const char* pix_type, const int fpsnum, const int fpsden,
              const char* index, const bool show, const int sarnum, const int sarden, ise_t* env);
    PVideoFrame __stdcall GetFrame(int n, ise_t *env);

    ~RawSource() { _close(fileHandle); }
    bool __stdcall GetParity(int n) { return vi.image_type == VideoInfo::IT_TFF; }
    void __stdcall GetAudio(void *buf, int64_t start, int64_t count, ise_t* env) {}
    const VideoInfo& __stdcall GetVideoInfo() { return vi; }
    int __stdcall SetCacheHints(int hints, int)
    {
        return hints == CACHE_GET_MTMODE ? MT_SERIALIZED : 0;
    }
};


void RawSource::setProcess(const char* pix_type)
{
    using std::make_tuple;
    constexpr int Y = PLANAR_Y, U = PLANAR_U, V = PLANAR_V;
    constexpr int G = PLANAR_G, B = PLANAR_B, R = PLANAR_R, A = PLANAR_A;
    constexpr int X = 99999999;

    typedef void (__stdcall *write_frame_t)(
        int, PVideoFrame&, uint8_t*, int*, int, ise_t*);
    typedef std::tuple<int, int, int, int, int, int, write_frame_t> pixel_format_t;

    std::unordered_map<std::string, pixel_format_t> table;

    table["BGR"]        = make_tuple(VideoInfo::CS_BGR24,       0, 1, 2, X, 1, write_planar);
    table["BGR24"]      = make_tuple(VideoInfo::CS_BGR24,       0, 1, 2, X, 1, write_planar);
    table["RGB"]        = make_tuple(VideoInfo::CS_BGR24,       2, 1, 0, X, 3, write_packed_reorder_8);
    table["RGB24"]      = make_tuple(VideoInfo::CS_BGR24,       2, 1, 0, X, 3, write_packed_reorder_8);

    table["BGR48"]      = make_tuple(VideoInfo::CS_BGR48,       0, 1, 2, X, 1, write_planar);
    table["RGB48"]      = make_tuple(VideoInfo::CS_BGR48,       2, 1, 0, X, 3, write_packed_reorder_16);

    table["BGRA"]       = make_tuple(VideoInfo::CS_BGR32,       0, 1, 2, 3, 1, write_planar);
    table["BGR32"]      = make_tuple(VideoInfo::CS_BGR32,       0, 1, 2, 3, 1, write_planar);
    table["RGBA"]       = make_tuple(VideoInfo::CS_BGR32,       2, 1, 0, 3, 4, write_packed_reorder_8);
    table["RGB32"]      = make_tuple(VideoInfo::CS_BGR32,       2, 1, 0, 3, 4, write_packed_reorder_8);
    table["ARGB"]       = make_tuple(VideoInfo::CS_BGR32,       3, 2, 1, 0, 4, write_packed_reorder_8);
    table["ARGB32"]     = make_tuple(VideoInfo::CS_BGR32,       3, 2, 1, 0, 4, write_packed_reorder_8);
    table["ABGR"]       = make_tuple(VideoInfo::CS_BGR32,       3, 0, 1, 2, 4, write_packed_reorder_8);
    table["ABGR32"]     = make_tuple(VideoInfo::CS_BGR32,       3, 0, 1, 2, 4, write_packed_reorder_8);

    table["BGR64"]      = make_tuple(VideoInfo::CS_BGR64,       0, 1, 2, 3, 1, write_planar);
    table["BGRA64"]     = make_tuple(VideoInfo::CS_BGR64,       0, 1, 2, 3, 1, write_planar);
    table["RGB64"]      = make_tuple(VideoInfo::CS_BGR64,       2, 1, 0, 3, 4, write_packed_reorder_16);
    table["RGBA64"]     = make_tuple(VideoInfo::CS_BGR64,       2, 1, 0, 3, 4, write_packed_reorder_16);
    table["ARGB64"]     = make_tuple(VideoInfo::CS_BGR64,       3, 2, 1, 0, 4, write_packed_reorder_16);
    table["ABGR64"]     = make_tuple(VideoInfo::CS_BGR64,       3, 0, 1, 2, 4, write_packed_reorder_16);

    table["YUY2"]       = make_tuple(VideoInfo::CS_YUY2,        0, 1, 2, 3, 1, write_planar);
    table["YUYV"]       = make_tuple(VideoInfo::CS_YUY2,        0, 1, 2, 3, 1, write_planar);
    table["UYVY"]       = make_tuple(VideoInfo::CS_YUY2,        1, 0, 3, 2, 4, write_packed_reorder_8);
    table["VYUY"]       = make_tuple(VideoInfo::CS_YUY2,        3, 0, 1, 2, 4, write_packed_reorder_8);

    table["GBRP"]       = make_tuple(VideoInfo::CS_RGBP,        G, B, R, X, 3, write_planar);
    table["GBRP10"]     = make_tuple(VideoInfo::CS_RGBP10,      G, B, R, X, 3, write_planar);
    table["GBRP12"]     = make_tuple(VideoInfo::CS_RGBP12,      G, B, R, X, 3, write_planar);
    table["GBRP14"]     = make_tuple(VideoInfo::CS_RGBP14,      G, B, R, X, 3, write_planar);
    table["GBRP16"]     = make_tuple(VideoInfo::CS_RGBP16,      G, B, R, X, 3, write_planar);
    table["GBRPS"]      = make_tuple(VideoInfo::CS_RGBPS,       G, B, R, X, 3, write_planar);

    table["GBRAP"]      = make_tuple(VideoInfo::CS_RGBAP,       G, B, R, A, 4, write_planar);
    table["GBRAP10"]    = make_tuple(VideoInfo::CS_RGBAP10,     G, B, R, A, 4, write_planar);
    table["GBRAP12"]    = make_tuple(VideoInfo::CS_RGBAP12,     G, B, R, A, 4, write_planar);
    table["GBRAP14"]    = make_tuple(VideoInfo::CS_RGBAP14,     G, B, R, A, 4, write_planar);
    table["GBRAP16"]    = make_tuple(VideoInfo::CS_RGBAP16,     G, B, R, A, 4, write_planar);
    table["GBRAPS"]     = make_tuple(VideoInfo::CS_RGBAPS,      G, B, R, A, 4, write_planar);

    table["YV24"]       = make_tuple(VideoInfo::CS_YV24,        Y, V, U, X, 3, write_planar);
    table["YUV444P8"]   = make_tuple(VideoInfo::CS_YV24,        Y, U, V, X, 3, write_planar);
    table["YUV444P9"]   = make_tuple(VideoInfo::CS_YUV444P10,   Y, U, V, X, 3, write_planar_9);
    table["YUV444P10"]  = make_tuple(VideoInfo::CS_YUV444P10,   Y, U, V, X, 3, write_planar);
    table["YUV444P12"]  = make_tuple(VideoInfo::CS_YUV444P12,   Y, U, V, X, 3, write_planar);
    table["YUV444P14"]  = make_tuple(VideoInfo::CS_YUV444P14,   Y, U, V, X, 3, write_planar);
    table["YUV444P16"]  = make_tuple(VideoInfo::CS_YUV444P16,   Y, U, V, X, 3, write_planar);
    table["YUV444PS"]   = make_tuple(VideoInfo::CS_YUV444PS,    Y, U, V, X, 3, write_planar);

    table["YUVA444P8"]  = make_tuple(VideoInfo::CS_YUVA444,     Y, U, V, A, 4, write_planar);
    table["YUVA444P10"] = make_tuple(VideoInfo::CS_YUV444P10,   Y, U, V, A, 4, write_planar);
    table["YUVA444P12"] = make_tuple(VideoInfo::CS_YUV444P12,   Y, U, V, A, 4, write_planar);
    table["YUVA444P14"] = make_tuple(VideoInfo::CS_YUV444P14,   Y, U, V, A, 4, write_planar);
    table["YUVA444P16"] = make_tuple(VideoInfo::CS_YUV444P16,   Y, U, V, A, 4, write_planar);
    table["YUVA444PS"]  = make_tuple(VideoInfo::CS_YUV444PS,    Y, U, V, A, 4, write_planar);

    table["YV16"]       = make_tuple(VideoInfo::CS_YV16,        Y, V, U, X, 3, write_planar);
    table["YUV422P8"]   = make_tuple(VideoInfo::CS_YV16,        Y, U, V, X, 3, write_planar);
    table["YUV422P9"]   = make_tuple(VideoInfo::CS_YUV444P10,   Y, U, V, X, 3, write_planar_9);
    table["YUV422P10"]  = make_tuple(VideoInfo::CS_YUV422P10,   Y, U, V, X, 3, write_planar);
    table["YUV422P12"]  = make_tuple(VideoInfo::CS_YUV422P12,   Y, U, V, X, 3, write_planar);
    table["YUV422P14"]  = make_tuple(VideoInfo::CS_YUV422P14,   Y, U, V, X, 3, write_planar);
    table["YUV422P16"]  = make_tuple(VideoInfo::CS_YUV422P16,   Y, U, V, X, 3, write_planar);
    table["YUV422PS"]   = make_tuple(VideoInfo::CS_YUV422PS,    Y, U, V, X, 3, write_planar);

    table["YUVA422P8"]  = make_tuple(VideoInfo::CS_YUVA422,     Y, U, V, A, 4, write_planar);
    table["YUVA422P10"] = make_tuple(VideoInfo::CS_YUVA422P10,  Y, U, V, A, 4, write_planar);
    table["YUVA422P12"] = make_tuple(VideoInfo::CS_YUVA422P12,  Y, U, V, A, 4, write_planar);
    table["YUVA422P14"] = make_tuple(VideoInfo::CS_YUVA422P14,  Y, U, V, A, 4, write_planar);
    table["YUVA422P16"] = make_tuple(VideoInfo::CS_YUVA422P16,  Y, U, V, A, 4, write_planar);
    table["YUVA422PS"]  = make_tuple(VideoInfo::CS_YUVA422PS,   Y, U, V, A, 4, write_planar);

    table["P210"]       = make_tuple(VideoInfo::CS_YUV422P10,   Y, U, V, X, 2, write_packed_chroma_16);
    table["P216"]       = make_tuple(VideoInfo::CS_YUV422P16,   Y, U, V, X, 2, write_packed_chroma_16);

    table["YV411"]      = make_tuple(VideoInfo::CS_YV411,       Y, V, U, X, 3, write_planar);
    table["Y41B"]       = make_tuple(VideoInfo::CS_YV411,       Y, V, U, X, 3, write_planar);
    table["YUV411P8"]   = make_tuple(VideoInfo::CS_YV411,       Y, U, V, X, 3, write_planar);

    table["YV12"]       = make_tuple(VideoInfo::CS_YV12,        Y, V, U, X, 3, write_planar);
    table["I420"]       = make_tuple(VideoInfo::CS_I420,        Y, U, V, X, 3, write_planar);
    table["IYUV"]       = make_tuple(VideoInfo::CS_I420,        Y, U, V, X, 3, write_planar);
    table["YUV420P8"]   = make_tuple(VideoInfo::CS_I420,        Y, U, V, X, 3, write_planar);
    table["YUV420P9"]   = make_tuple(VideoInfo::CS_YUV420P10,   Y, U, V, X, 3, write_planar_9);
    table["YUV420P10"]  = make_tuple(VideoInfo::CS_YUV420P10,   Y, U, V, X, 3, write_planar);
    table["YUV420P12"]  = make_tuple(VideoInfo::CS_YUV420P12,   Y, U, V, X, 3, write_planar);
    table["YUV420P14"]  = make_tuple(VideoInfo::CS_YUV420P14,   Y, U, V, X, 3, write_planar);
    table["YUV420P16"]  = make_tuple(VideoInfo::CS_YUV420P16,   Y, U, V, X, 3, write_planar);
    table["YUV420PS"]   = make_tuple(VideoInfo::CS_YUV420PS,    Y, U, V, X, 3, write_planar);

    table["YUVA420P8"]  = make_tuple(VideoInfo::CS_YUVA420,     Y, U, V, A, 4, write_planar);
    table["YUVA420P10"] = make_tuple(VideoInfo::CS_YUVA420P10,  Y, U, V, A, 4, write_planar);
    table["YUVA420P12"] = make_tuple(VideoInfo::CS_YUVA420P12,  Y, U, V, A, 4, write_planar);
    table["YUVA420P14"] = make_tuple(VideoInfo::CS_YUVA420P14,  Y, U, V, A, 4, write_planar);
    table["YUVA420P16"] = make_tuple(VideoInfo::CS_YUVA420P16,  Y, U, V, A, 4, write_planar);
    table["YUVA420PS"]  = make_tuple(VideoInfo::CS_YUVA420PS,   Y, U, V, A, 4, write_planar);

    table["NV12"]       = make_tuple(VideoInfo::CS_I420,        Y, U, V, X, 2, write_packed_chroma_8);
    table["NV21"]       = make_tuple(VideoInfo::CS_YV12,        Y, V, U, X, 2, write_packed_chroma_8);
    table["P010"]       = make_tuple(VideoInfo::CS_YUV420P10,   Y, U, V, X, 2, write_packed_chroma_16);
    table["P016"]       = make_tuple(VideoInfo::CS_YUV420P16,   Y, U, V, X, 2, write_packed_chroma_16);

    table["Y8"]         = make_tuple(VideoInfo::CS_Y8,          Y, X, X, X, 1, write_planar);
    table["GREY8"]      = make_tuple(VideoInfo::CS_Y8,          Y, X, X, X, 1, write_planar);
    table["Y10"]        = make_tuple(VideoInfo::CS_Y10,         Y, X, X, X, 1, write_planar);
    table["Y12"]        = make_tuple(VideoInfo::CS_Y12,         Y, X, X, X, 1, write_planar);
    table["Y14"]        = make_tuple(VideoInfo::CS_Y14,         Y, X, X, X, 1, write_planar);
    table["Y16"]        = make_tuple(VideoInfo::CS_Y16,         Y, X, X, X, 1, write_planar);
    table["GREY16"]     = make_tuple(VideoInfo::CS_Y16,         Y, X, X, X, 1, write_planar);
    table["Y32"]        = make_tuple(VideoInfo::CS_Y32,         Y, X, X, X, 1, write_planar);
    table["GREYS"]      = make_tuple(VideoInfo::CS_Y32,         Y, X, X, X, 1, write_planar);

    auto key = std::string(pix_type);
    std::transform(key.begin(), key.end(), key.begin(), ::toupper);

    std::tie(vi.pixel_type, order[0], order[1], order[2], order[3], col_count, writeDestFrame) = table[key];

    validate(vi.pixel_type == 0, "Invalid pixel type. Supported types are bellows.:\n"
        "RGB, RGBA, BGR, BGRA, ARGB, ABGR, RGB48, BGR48, RGBA64, BGRA64, ARGB64, ABGR64,\n"
        "GBRP, GBRP10, GBRP12, GBRP14, GBRP16, GBRPS, GBRAP, GBRAP10, GBRAP12, GBRAP14, GBRAP16, GBRAPS,\n"
        "Y8, GREY8, Y10, Y12, Y14, Y16, GREY16, Y32, GREYS, YUY2, YUYV, UYVY, YVYU, VYUY,\n"
        "YV24, I444, YUV444P8, YUV444P9, YUV444P10, YUV444P12, YUV444P14, YUV444P16, YUV444PS,\n"
        "YUVA444P8, YUVA444P10, YUVA444P12, YUVA444P14, YUVA444P16, YUVA444PS,\n"
        "YV16, I422, YUV422P8, YUV422P9, YUV422P10, YUV422P12, YUV422P14, YUV422P16, YUV422PS,\n"
        "YUVA422P8, YUVA422P10, YUVA422P12, YUVA422P14, YUVA422P16, YUVA422PS,\n"
        "YV12, I420, IYUV, YUV420P8, YUV420P9, YUV420P10, YUV420P12, YUV420P14, YUV420P16, YUV420PS,\n"
        "YUVA420P8, YUVA420P10, YUVA420P12, YUVA420P14, YUVA420P16, YUVA420PS,\n"
        "YV411, Y41B, NV12, NV21, P210, P216, P010, P016");

}


RawSource::RawSource(const char *source, const int width, const int height,
                     const char *ptype, const int fpsnum, const int fpsden,
                     const char *a_index, const bool s, const int sarnum, const int sarden, ise_t* env) : show(s), sar_num(sarnum), sar_den(sarden)
{
#ifdef _WIN32
    _sopen_s(&fileHandle, source, _O_BINARY | _O_RDONLY, _SH_DENYWR, 0);
    validate(fileHandle == -1, "Cannot open videofile.");

    fileSize = _filelengthi64(fileHandle);
    validate(fileSize == -1L, "Cannot get videofile length.");
#else
    struct flock fl;
    struct stat buf;
    fileHandle = open(source, O_RDONLY);
    validate(fileHandle == -1, "Cannot open videofile.");
    fl.l_type = F_RDLCK;
    fl.l_whence = SEEK_SET;
    fl.l_start = 0;
    fl.l_len = 0;
    validate((fcntl(fileHandle, F_SETLKW, &fl) == -1), "Cannot lock videofile.");

    validate((fstat(fileHandle, &buf) != 0), "Cannot get videofile length.");
    fileSize = buf.st_size;
#endif

    memset(&vi, 0, sizeof(VideoInfo));
    vi.width = width;
    vi.height = height;
    vi.SetFPS(fpsnum, fpsden);
    vi.SetFieldBased(false);

    char pix_type[16] = {};
    strcpy_s(pix_type, 16, ptype);

    int64_t header_offset = 0;
    int64_t frame_offset = 0;

    if (strlen(a_index) == 0) { //use header if valid else width, height, pixel_type from AVS are used
        std::vector<char> read_buff(256, 0);
        char* data = read_buff.data();
        _read(fileHandle, data, read_buff.size()); //read some bytes and test on header
        if (parse_y4m(read_buff, vi, header_offset, frame_offset, sar_den, sar_num)) {
            strcpy_s(pix_type, 16, data);
        }
    }

    setProcess(pix_type);

    size_t framesize = vi.width * vi.height * vi.BitsPerPixel() / 8;

    int maxframe = static_cast<int>(fileSize / framesize);    //1 = one frame

    validate(maxframe < 1, "File too small for even one frame.");

    //index build using string descriptor
#ifdef __GNUC__
    std::vector<struct rindex> rawindex;
#else
    std::vector<rindex> rawindex;
#endif
    set_rawindex(rawindex, a_index, header_offset, frame_offset, framesize);

    has_at_least_v8 = true;
    try { env->CheckVersion(8); }
    catch (const AvisynthError&) { has_at_least_v8 = false; }

    //create full index and get number of frames.
    if (has_at_least_v8)
    {
        auto free_buffer = [](void* p, ise_t* e) {
            static_cast<IScriptEnvironment*>(e)->Free(p);
            p = nullptr;
        };

        void* b = env->Allocate((static_cast<int64_t>(maxframe) + 1) * sizeof(i_struct), 8, AVS_NORMAL_ALLOC);
        validate(!b, "failed to allocate index array.");
        env->AtExit(free_buffer, b);
        index = reinterpret_cast<i_struct*>(b);
        vi.num_frames = generate_index(index, rawindex, framesize, fileSize);

        b = env->Allocate(vi.BytesFromPixels(vi.width * vi.height), 64, AVS_NORMAL_ALLOC);
        validate(!b, "failed to allocate read buffer.");
        env->AtExit(free_buffer, b);
        rawbuf = reinterpret_cast<uint8_t*>(b);
    }
    else
    {
        auto free_buffer = [](void* p, ise_t* e) {
            _aligned_free(p);
            p = nullptr;
        };

        void* b = _aligned_malloc((static_cast<int64_t>(maxframe) + 1) * sizeof(i_struct), 8);
        validate(!b, "failed to allocate index array.");
        env->AtExit(free_buffer, b);
        index = reinterpret_cast<i_struct*>(b);
        vi.num_frames = generate_index(index, rawindex, framesize, fileSize);

        b = _aligned_malloc(vi.BytesFromPixels(vi.width * vi.height), 64);
        validate(!b, "failed to allocate read buffer.");
        env->AtExit(free_buffer, b);
        rawbuf = reinterpret_cast<uint8_t*>(b);
    }

    env->SetVar(env->Sprintf("%s", "FFSAR_NUM"), sar_num);
    env->SetVar(env->Sprintf("%s", "FFSAR_DEN"), sar_den);
    if (sar_num > 0 && sar_den > 0)
        env->SetVar(env->Sprintf("%s", "FFSAR"), sar_num / static_cast<double>(sar_den));
}


PVideoFrame __stdcall RawSource::GetFrame(int n, ise_t* env)
{
    auto dst = env->NewVideoFrame(vi);

    if (_lseeki64(fileHandle, index[n].index, SEEK_SET) == -1L) {
        // black frame with message
        write_black_frame(dst, vi);
        env->ApplyMessage(&dst, vi, "failed to seek file!", vi.width,
                          0xFFFFFF, 0xFFFFFF, 0);
        return dst;
    }

    writeDestFrame(fileHandle, dst, rawbuf, order, col_count, env);

    if (show) { //output debug info
        char info[64];
        sprintf(info, "%d : %" PRIi64 " %c", n, index[n].index, index[n].type);
        env->ApplyMessage(&dst, vi, info, vi.width / 2, 0xFFFFFF, 0, 0);
    }

    if (has_at_least_v8)
    {
        AVSMap* props = env->getFramePropsRW(dst);
        env->propSetInt(props, "_DurationNum", vi.fps_denominator, 0);
        env->propSetInt(props, "_DurationDen", vi.fps_numerator, 0);
        env->propSetInt(props, "_SARNum", sar_num, 0);
        env->propSetInt(props, "_SARDen", sar_den, 0);
    }

    return dst;
}


AVSValue __cdecl create_rawsource(AVSValue args, void* user_data, ise_t* env)
{
    char buff[128] = {};

    try {
        validate(!args[0].Defined(), "No source specified");

        const char *source = args[0].AsString();
        const int width = args[1].AsInt(720);
        const int height = args[2].AsInt(576);
        const char *pix_type = args[3].AsString("YUV420P8");
        const int fpsnum = args[4].AsInt(25);
        const int fpsden = args[5].AsInt(1);
        const char *index = args[6].AsString("");
        const bool show = args[7].AsBool(false);
        const int sarnum = args[8].AsInt(0);
        const int sarden = args[9].AsInt(0);

        if (width < MIN_WIDTH || height < MIN_HEIGHT) {
            snprintf(buff, 127, "width and height need to be %u x %u or higher.",
                     MIN_WIDTH, MIN_HEIGHT);
            throw std::runtime_error(buff);
        }

        validate(strlen(pix_type) > 15, "pixel_type is too long.");
        validate(fpsnum < 1 || fpsden < 1, "fpsnum and fpsden need to be 1 or higher.");
        validate(sarnum < 0 || sarden < 0, "sarnum and sarden need to be 0 or higher.");

        return new RawSource(
            source,
            width,
            height,
            pix_type,
            fpsnum,
            fpsden,                   
            index,
            show,
            sarnum,
            sarden,
            env);

    } catch (std::runtime_error& e) {
        env->ThrowError("RawSourcePlus: %s", e.what());
    }

    return 0;
}

const AVS_Linkage* AVS_linkage = nullptr;

extern "C" __declspec(dllexport) const char* __stdcall
AvisynthPluginInit3(ise_t* env, const AVS_Linkage* const vectors)
{
    AVS_linkage = vectors;

    const char* args =
        "[file]s"
        "[width]i"
        "[height]i"
        "[pixel_type]s"
        "[fpsnum]i"
        "[fpsden]i"
        "[index]s"
        "[show]b"
        "[sarnum]i"
        "[sarden]i";

    env->AddFunction("RawSourcePlus", args, create_rawsource, nullptr);

    return "RawSource for Avisynth+.";
}

